File: analyze_comments.py
#!/usr/bin/env python3
"""
Analyze multi-line comment blocks in card.js and generate compression mapping.
"""

import re
import json

def analyze_comment_blocks(filepath):
    """Find and analyze all multi-line comment blocks (2+ consecutive lines)."""

    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    blocks = []
    current_block = None

    for i, line in enumerate(lines, 1):
        # Check if line is a comment (starting with //)
        stripped = line.lstrip()
        if stripped.startswith('//'):
            comment_text = stripped[2:].strip()

            if current_block is None:
                # Start new block
                current_block = {
                    'startLine': i,
                    'endLine': i,
                    'lines': [comment_text]
                }
            else:
                # Continue block
                current_block['endLine'] = i
                current_block['lines'].append(comment_text)
        else:
            # Non-comment line - finish current block if it exists
            if current_block is not None:
                # Only keep blocks with 2+ lines
                if len(current_block['lines']) >= 2:
                    blocks.append(current_block)
                current_block = None

    # Handle case where file ends with a comment block
    if current_block is not None and len(current_block['lines']) >= 2:
        blocks.append(current_block)

    return blocks

def compress_block(block):
    """Generate semantic compression for a comment block."""
    lines = block['lines']
    full_text = ' '.join(lines)

    # Pattern-based semantic compression
    compressions = {
        # Phase/Section markers
        r'={3,}.*PHASE.*={3,}': lambda m: '// ' + re.search(r'PHASE[^=]*', full_text).group().strip(),
        r'={3,}': lambda m: '// ' + lines[0] if lines else '// Section marker',
        r'/{3,}': lambda m: '// ' + lines[0] if lines else '// Section divider',

        # State machine related
        r'STATE MACHINE': lambda m: '// STATE MACHINE: ' + (lines[1] if len(lines) > 1 else lines[0]),
        r'TRANSITION': lambda m: '// State transition: ' + full_text[:80],

        # Fix/Bug related
        r'FIX:|FIXED:|BUG:': lambda m: '// FIX: ' + full_text[:100].replace('FIX:', '').replace('FIXED:', '').replace('BUG:', '').strip(),

        # Performance/Optimization
        r'PERFORMANCE|OPTIMIZATION': lambda m: '// PERF: ' + full_text[:100],

        # Security
        r'SECURITY|XSS|SANITIZE': lambda m: '// SECURITY: ' + full_text[:100],

        # Defensive/Critical
        r'CRITICAL:|DEFENSIVE:|IMPORTANT:': lambda m: '// ' + full_text[:120],

        # Accessibility
        r'ACCESSIBILITY|ARIA|screen reader': lambda m: '// A11Y: ' + full_text[:100],

        # MO/M3 markers (optimization work)
        r'M3:|MO\d+:': lambda m: '// ' + full_text[:120],

        # TODO/NOTE/WARNING
        r'TODO:|NOTE:|WARNING:': lambda m: '// ' + full_text[:100],

        # Helper function descriptions
        r'Helper function': lambda m: '// Helper: ' + full_text[:80],

        # Scenario descriptions
        r'Scenario \d+': lambda m: '// ' + full_text[:100],
    }

    # Try pattern matching
    for pattern, compress_fn in compressions.items():
        if re.search(pattern, full_text, re.IGNORECASE):
            return compress_fn(None)

    # Default compression strategies based on content analysis

    # Check if it's explaining what something does
    if any(keyword in full_text.lower() for keyword in ['this ', 'we ', 'note that', 'function', 'called']):
        # Extract key action/purpose
        if len(full_text) <= 120:
            return '// ' + full_text
        else:
            # Take first substantial sentence
            sentences = re.split(r'[.!?]\s+', full_text)
            if sentences:
                return '// ' + sentences[0][:120]

    # Check if it's configuration/settings
    if any(keyword in full_text.lower() for keyword in ['setting', 'config', 'option', 'parameter']):
        return '// Config: ' + full_text[:100]

    # Check if it's describing logic flow
    if any(keyword in full_text.lower() for keyword in ['if ', 'when ', 'check ', 'handle']):
        return '// Logic: ' + full_text[:100]

    # Default: take first line if short enough, otherwise summarize
    if len(lines[0]) <= 100:
        return '// ' + lines[0]
    else:
        return '// ' + full_text[:120]

def main():
    filepath = r'c:\Users\ppavl\OneDrive\Active projects\mofacts\mofacts\client\views\experiment\card.js'

    print("Analyzing comment blocks in card.js...")
    blocks = analyze_comment_blocks(filepath)

    print(f"Found {len(blocks)} multi-line comment blocks (2+ consecutive lines)")

    # Generate compression mapping
    result = []
    for i, block in enumerate(blocks, 1):
        compressed = compress_block(block)
        result.append({
            'blockNum': i,
            'startLine': block['startLine'],
            'endLine': block['endLine'],
            'lineCount': len(block['lines']),
            'preview': ' '.join(block['lines'][:2])[:100] + ('...' if len(' '.join(block['lines'][:2])) > 100 else ''),
            'compressed': compressed
        })

    # Output JSON
    output_file = r'c:\Users\ppavl\OneDrive\Active projects\mofacts\comment_compression_mapping.json'
    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(result, f, indent=2, ensure_ascii=False)

    print(f"\nGenerated compression mapping saved to: {output_file}")
    print(f"Total blocks: {len(result)}")

    # Print summary
    print("\n=== SAMPLE OUTPUT (first 5 blocks) ===")
    for item in result[:5]:
        print(f"\nBlock {item['blockNum']} (lines {item['startLine']}-{item['endLine']}, {item['lineCount']} lines)")
        print(f"Preview: {item['preview']}")
        print(f"Compressed: {item['compressed']}")

    print(f"\n=== BLOCKS 21-30 (continuing from previous analysis) ===")
    for item in result[20:30]:
        print(f"\nBlock {item['blockNum']} (lines {item['startLine']}-{item['endLine']}, {item['lineCount']} lines)")
        print(f"Preview: {item['preview']}")
        print(f"Compressed: {item['compressed']}")

if __name__ == '__main__':
    main()

File: scripts/analyze_comments.js
#!/usr/bin/env node
/**
 * Analyze multi-line comment blocks in card.js and generate compression mapping.
 */

const fs = require('fs');
const path = require('path');

function analyzeCommentBlocks(filepath) {
  const content = fs.readFileSync(filepath, 'utf-8');
  const lines = content.split('\n');

  const blocks = [];
  let currentBlock = null;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const stripped = line.trimStart();

    if (stripped.startsWith('//')) {
      const commentText = stripped.substring(2).trim();

      if (currentBlock === null) {
        // Start new block
        currentBlock = {
          startLine: i + 1,
          endLine: i + 1,
          lines: [commentText]
        };
      } else {
        // Continue block
        currentBlock.endLine = i + 1;
        currentBlock.lines.push(commentText);
      }
    } else {
      // Non-comment line - finish current block if it exists
      if (currentBlock !== null) {
        // Only keep blocks with 2+ lines
        if (currentBlock.lines.length >= 2) {
          blocks.push(currentBlock);
        }
        currentBlock = null;
      }
    }
  }

  // Handle case where file ends with a comment block
  if (currentBlock !== null && currentBlock.lines.length >= 2) {
    blocks.push(currentBlock);
  }

  return blocks;
}

function compressBlock(block) {
  const lines = block.lines;
  const fullText = lines.join(' ');

  // Pattern-based semantic compression
  const patterns = [
    // Phase/Section markers
    { pattern: /={3,}.*PHASE.*={3,}/i, fn: () => {
      const match = fullText.match(/PHASE[^=]*/i);
      return '// ' + (match ? match[0].trim() : lines[0]);
    }},
    { pattern: /={3,}/, fn: () => '// ' + lines[0] },
    { pattern: /\/{3,}/, fn: () => '// ' + lines[0] },

    // State machine
    { pattern: /STATE MACHINE/i, fn: () => '// STATE MACHINE: ' + (lines[1] || lines[0]) },
    { pattern: /TRANSITION/i, fn: () => '// State transition: ' + fullText.substring(0, 80) },

    // Fix/Bug
    { pattern: /\b(FIX|FIXED|BUG):/i, fn: () => {
      return '// FIX: ' + fullText.substring(0, 100).replace(/FIX:|FIXED:|BUG:/gi, '').trim();
    }},

    // Performance
    { pattern: /\b(PERFORMANCE|OPTIMIZATION)\b/i, fn: () => '// PERF: ' + fullText.substring(0, 100) },

    // Security
    { pattern: /\b(SECURITY|XSS|SANITIZE)\b/i, fn: () => '// SECURITY: ' + fullText.substring(0, 100) },

    // Defensive/Critical
    { pattern: /\b(CRITICAL|DEFENSIVE|IMPORTANT):/i, fn: () => '// ' + fullText.substring(0, 120) },

    // Accessibility
    { pattern: /\b(ACCESSIBILITY|ARIA|screen reader)\b/i, fn: () => '// A11Y: ' + fullText.substring(0, 100) },

    // MO/M3 markers
    { pattern: /\b(M3|MO\d+):/i, fn: () => '// ' + fullText.substring(0, 120) },

    // TODO/NOTE/WARNING
    { pattern: /\b(TODO|NOTE|WARNING):/i, fn: () => '// ' + fullText.substring(0, 100) },

    // Helper function
    { pattern: /Helper function/i, fn: () => '// Helper: ' + fullText.substring(0, 80) },

    // Scenario
    { pattern: /Scenario \d+/i, fn: () => '// ' + fullText.substring(0, 100) },
  ];

  // Try pattern matching
  for (const {pattern, fn} of patterns) {
    if (pattern.test(fullText)) {
      return fn();
    }
  }

  // Default compression strategies

  // Check if explaining purpose
  const purposeKeywords = ['this ', 'we ', 'note that', 'function', 'called'];
  if (purposeKeywords.some(kw => fullText.toLowerCase().includes(kw))) {
    if (fullText.length <= 120) {
      return '// ' + fullText;
    } else {
      const sentences = fullText.split(/[.!?]\s+/);
      if (sentences.length > 0) {
        return '// ' + sentences[0].substring(0, 120);
      }
    }
  }

  // Configuration
  const configKeywords = ['setting', 'config', 'option', 'parameter'];
  if (configKeywords.some(kw => fullText.toLowerCase().includes(kw))) {
    return '// Config: ' + fullText.substring(0, 100);
  }

  // Logic flow
  const logicKeywords = ['if ', 'when ', 'check ', 'handle'];
  if (logicKeywords.some(kw => fullText.toLowerCase().includes(kw))) {
    return '// Logic: ' + fullText.substring(0, 100);
  }

  // Default
  if (lines[0].length <= 100) {
    return '// ' + lines[0];
  } else {
    return '// ' + fullText.substring(0, 120);
  }
}

function main() {
  const filepath = path.join(__dirname, 'mofacts', 'client', 'views', 'experiment', 'card.js');

  console.log('Analyzing comment blocks in card.js...');
  const blocks = analyzeCommentBlocks(filepath);

  console.log(`Found ${blocks.length} multi-line comment blocks (2+ consecutive lines)`);

  // Generate compression mapping
  const result = blocks.map((block, i) => {
    const preview = block.lines.slice(0, 2).join(' ');
    return {
      blockNum: i + 1,
      startLine: block.startLine,
      endLine: block.endLine,
      lineCount: block.lines.length,
      preview: preview.substring(0, 100) + (preview.length > 100 ? '...' : ''),
      compressed: compressBlock(block)
    };
  });

  // Output JSON
  const outputFile = path.join(__dirname, 'comment_compression_mapping.json');
  fs.writeFileSync(outputFile, JSON.stringify(result, null, 2), 'utf-8');

  console.log(`\nGenerated compression mapping saved to: ${outputFile}`);
  console.log(`Total blocks: ${result.length}`);

  // Print summary
  console.log('\n=== SAMPLE OUTPUT (first 5 blocks) ===');
  result.slice(0, 5).forEach(item => {
    console.log(`\nBlock ${item.blockNum} (lines ${item.startLine}-${item.endLine}, ${item.lineCount} lines)`);
    console.log(`Preview: ${item.preview}`);
    console.log(`Compressed: ${item.compressed}`);
  });

  console.log('\n=== BLOCKS 21-30 (continuing from previous analysis) ===');
  result.slice(20, 30).forEach(item => {
    console.log(`\nBlock ${item.blockNum} (lines ${item.startLine}-${item.endLine}, ${item.lineCount} lines)`);
    console.log(`Preview: ${item.preview}`);
    console.log(`Compressed: ${item.compressed}`);
  });
}

main();
